File: python/890. Find and Replace Pattern.py
class Solution(object):
    def findAndReplacePattern(self, words, pattern):
        """
        :type words: List[str]
        :type pattern: str
        :rtype: List[str]
        """
        def check(w):
            if len(w) != len(pattern):
                return False
            w2p = {}
            p2w = {}
            for i in range(len(w)):
                if w[i] not in w2p and pattern[i] not in p2w:
                    w2p[w[i]] = pattern[i]
                    p2w[pattern[i]] = w[i]
                elif w[i] in w2p and pattern[i] in p2w:
                    if w2p[w[i]] != pattern[i] or p2w[pattern[i]] != w[i]:
                        return False
                else:
                    return False
            return True
        
        return filter(check, words)
            

File: go/src/leetcode/890/890. Find and Replace Pattern.go
/**
@author ZhengHao Lou
Date    2022/06/12
*/
package main

import "fmt"

/**
https://leetcode.cn/problems/find-and-replace-pattern/
*/
func findAndReplacePattern(words []string, pattern string) []string {
	var ans []string
NEXT:
	for _, w := range words {
		mp := [2][26]byte{}
		n := len(w)
		for i := 0; i < n; i++ {
			ip, iw := index(pattern[i]), index(w[i])
			if mp[1][ip] != 0 && mp[0][iw] == 0 {
				continue NEXT
			}
			if mp[1][ip] == 0 && mp[0][iw] != 0 {
				continue NEXT
			}

			if (mp[1][ip] != 0 && mp[0][iw] != 0) && (mp[1][ip] != w[i] || mp[0][iw] != pattern[i]) {
				continue NEXT
			}
			mp[1][ip] = w[i]
			mp[0][iw] = pattern[i]
		}
		ans = append(ans, w)
	}
	return ans
}

func index(b byte) int {
	return int(b - 'a')
}

func main() {
	fmt.Println(findAndReplacePattern([]string{"abc", "deq", "mee", "aqq", "dkd", "ccc"}, "abb"))
}
